### PR TITLE
grainAllocation: inject optional mapping over cred scores

### PR DESCRIPTION
**Pull request by elihanover**
This commit allows injecting an optional `CredMapping :: CredScores => CredScores` into the receipt functions for each policy type (besides special by design).  In practice, we will use this for calculating grain distributions based on quadratically weighted cred; however, it lets us add any `CredScores` garnish more generally.  This makes policy building easier as we can tweak policies and their parameters on the fly without changing the underlying types.  This should also speed up any future cryptoeconomics research.

The obvious option is to create a new policy for each, but modifying policies does not scale will in this case.  Instead, injecting a more general mapping (which defaults to `id`) allows us to use the core logic of the core `Balanced`, `Immediate`, and `Recent` policies.

__Test Plan__
Here, I've added unit tests that test this by injecting a function to quadratically weight cred.

### CODE REVIEW

File: src/core/ledger/grainAllocation.js [hunk 1]
```diff
@@ -66,17 +66,20 @@ export type AllocationPolicy =
 export type BalancedPolicy = {|
   +policyType: Balanced,
   +budget: G.Grain,
+  +mapping?: Function,
```
Comment (hammadj) [inline src/core/ledger/grainAllocation.js:69]: Should add type information for the inputs / outputs of this function for typesafety

File: src/core/ledger/grainAllocation.js [hunk 2]
```diff
@@ -66,17 +66,20 @@ export type AllocationPolicy =
...
 |};
 
 export type ImmediatePolicy = {|
   +policyType: Immediate,
   +budget: G.Grain,
+  +mapping?: Function,
```
Comment (hammadj) [inline src/core/ledger/grainAllocation.js:75]: The name "mapping" isn't 100% clear, leads me to believe that this is a mapping for grain amounts, but seems like its actually a `credMapping`

File: src/core/ledger/grainAllocation.js [hunk 3]
```diff
@@ -214,12 +222,13 @@ function receipts(
...
+  identities: ProcessedIdentities,
+  mapping?: Function = (id) => id
 ): $ReadOnlyArray<GrainReceipt> {
-  const amounts = G.splitBudget(
-    budget,
-    identities.map((i) => i.mostRecentCred)
-  );
+  const credPerIdentity = identities.map((i) => i.mostRecentCred);
+  // if (!mapping) mapping = (id) => id
+  const modifiedCred = credPerIdentity.map(mapping);
```
Comment (hammadj) [inline src/core/ledger/grainAllocation.js:230]: I think the mapping function should accept the entire array of cred scores instead of each individual cred score. That way it can have context for what the entire range of values are and allow for more advanced mapping strategies (e.g. bell curve)

File: src/core/ledger/grain.js [hunk 4]
```diff
@@ -60,6 +60,9 @@ export function mul(a: Grain, b: Grain): Grain {
 export function div(a: Grain, b: Grain): Grain {
   return bigInt(a).divide(bigInt(b)).toString();
 }
+export function pow(a: Grain, b: Grain): Grain {
+  return bigInt(a).pow(bigInt(b)).toString();
+}
```
Comment (decentralion) [inline src/core/ledger/grain.js:65]: This doesn't seem right. Why would we be raising something to a `Grain` power? As an example, let `one` be one grain. Then `Grain.power(one, one)` is `(10^18) ^ (10^18)`. Also this forbids doing non-integer powers which frankly are the powers I"m most interested in (e.g. sqrt(x) is pow(x, 0.5)).

Also: if adding a new mathematical primitive to the grain module, please make that into its own PR.

File: src/core/ledger/grainAllocation.js [hunk 5]
```diff
@@ -66,17 +66,20 @@ export type AllocationPolicy =
...
   +policyType: Immediate,
   +budget: G.Grain,
+  +credMapping?: CredMapping,
 |};
 
 export type RecentPolicy = {|
   +policyType: Recent,
   +budget: G.Grain,
   +discount: Discount,
+  +credMapping?: CredMapping,
```
Comment (decentralion) [inline src/core/ledger/grainAllocation.js:82]: This is a no-go. The Policies need to be serializable (they are included in the ledger event), but functions are not serializable.

**Review by decentralion — CHANGES_REQUESTED** [review]
Requesting changes because we can't have arbitrary functions in the Policy types, because those types all get serialized inside the ledger, and functions can't be serialized. Need to take a different approach.

File: src/core/ledger/grainAllocation.js (hunk 5 above)
Comment (elihanover) [reply src/core/ledger/grainAllocation.js:82]: Ok, you had mentioned this previously but I had forgotten.  I'll close this PR.
